### PR TITLE
refactor: decompose the autofocus directive by handlers

### PR DESCRIPTION
**Pull request by splincode**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows [Conventional Commits](https://www.conventionalcommits.org/en/)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [x] Refactoring
- [ ] Code style update
- [ ] Build or CI related changes
- [ ] Documentation content changes

## What is the current behavior?

Part of #1634

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: projects/cdk/directives/auto-focus/handlers/abstract.handler.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import {ElementRef} from '@angular/core';
+import {TuiFocusableElementAccessor, TuiNativeFocusableElement} from '@taiga-ui/cdk';
```
Comment (nsbarsukov) [inline projects/cdk/directives/auto-focus/handlers/abstract.handler.ts:2]: fix import

Comment (splincode) [reply projects/cdk/directives/auto-focus/handlers/abstract.handler.ts]: fixed

File: projects/cdk/directives/auto-focus/autofocus.directive.ts [hunk 2]
```diff
@@ -10,25 +10,31 @@ import {
     Self,
 } from '@angular/core';
 import {ANIMATION_FRAME} from '@ng-web-apis/common';
-import {POLLING_TIME} from '@taiga-ui/cdk/constants';
-import {
-    TuiFocusableElementAccessor,
-    TuiNativeFocusableElement,
-} from '@taiga-ui/cdk/interfaces';
+import {TuiIosAutofocusHandler} from '@taiga-ui/cdk/directives';
```
[the remark this answers is not in the document]
Comment (splincode) [reply projects/cdk/directives/auto-focus/autofocus.directive.ts]: fixed

**Review by nsbarsukov — APPROVED** [review]

File: projects/cdk/directives/auto-focus/autofocus.directive.ts [hunk 3]
```diff
@@ -10,25 +10,31 @@ import {
...
 export class TuiAutoFocusDirective implements AfterViewInit {
+    private readonly handler: TuiAutofocusHandler = this.isIos
+        ? new TuiIosAutofocusHandler(
+              this.tuiFocusableComponent,
+              this.elementRef,
+              this.renderer,
+              this.ngZone,
+          )
+        : new TuiDefaultAutofocusHandler(
+              this.tuiFocusableComponent,
```
Comment (waterplea) [inline projects/cdk/directives/auto-focus/autofocus.directive.ts:33]: I wish there was a way to use `useClass` so we don't have to know all the dependencies. Maybe just for this situation we can reach for global variable? Something like:
```ts
{
  provider: AbstractTuiAutofocusHandler,
  useClass: typeof navigator !== 'undefined' && isApple(navigator)
    ? TuiIosAutofocusHandler
    : TuiDefaultAutofocusHandler
}
```

**Review by waterplea — CHANGES_REQUESTED** [review]

File: projects/cdk/directives/auto-focus/autofocus.directive.ts (hunk 3 above)
Comment (splincode) [reply projects/cdk/directives/auto-focus/autofocus.directive.ts:33]: @waterplea sorry, I have this
image

Comment (splincode) [reply projects/cdk/directives/auto-focus/autofocus.directive.ts:33]: @waterplea sorry, I have this
image

File: projects/cdk/directives/auto-focus/handlers/default.handler.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import {TuiFocusableElementAccessor} from '@taiga-ui/cdk/interfaces';
+import {Observable, race, timer} from 'rxjs';
+import {map, skipWhile, take, throttleTime} from 'rxjs/operators';
+
+import {AbstractTuiAutofocusHandler} from './abstract.handler';
+
+const TIMEOUT = 1000;
+const NG_ANIMATION_SELECTOR = '.ng-animating';
+
+export class TuiDefaultAutofocusHandler extends AbstractTuiAutofocusHandler {
```
[the remark this answers is not in the document]
Comment (splincode) [reply projects/cdk/directives/auto-focus/handlers/default.handler.ts:12]: TuiFocusableElementAccessor, ElementRef cannot inject to service which marked by `Injectable`

[the remark this answers is not in the document]
Comment (splincode) [reply projects/cdk/directives/auto-focus/handlers/default.handler.ts:12]: TuiFocusableElementAccessor, ElementRef cannot inject to service which marked by `Injectable`

File: projects/cdk/tokens/is-apple.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+import {TUI_IS_IOS} from './is-ios';
+
+const SAFARI_REG_EXP = /^((?!chrome|android).)*safari/i;
+
+export const TUI_IS_APPLE = new InjectionToken<boolean>(
+    'Apple(safari/webkit) detection',
+    {
+        factory: () =>
+            inject(TUI_IS_IOS) ||
+            SAFARI_REG_EXP.test(inject(USER_AGENT).toLowerCase() || ''),
```
Comment (waterplea) [inline projects/cdk/tokens/is-apple.ts:13]: Why fallback to an empty string?

File: projects/cdk/directives/auto-focus/autofocus.directive.ts [hunk 6]
```diff
@@ -10,25 +10,31 @@ import {
...
+import {TuiIosAutofocusHandler} from './handlers/ios.handler';
 
 // TODO: 3.0 change input name to tuiAutoFocus and handle empty string
 // TODO: refactor on this whole thing in 3.0
-@Directive({
-    selector: '[tuiAutoFocus]',
-})
+@Directive({selector: '[tuiAutoFocus]'})
 export class TuiAutoFocusDirective implements AfterViewInit {
+    private readonly handler: TuiAutofocusHandler = this.isIos
```
[the remark this answers is not in the document]
Comment (splincode) [reply projects/cdk/directives/auto-focus/autofocus.directive.ts]: fixed

File: projects/cdk/directives/auto-focus/handlers/default.handler.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (splincode) [reply projects/cdk/directives/auto-focus/handlers/default.handler.ts:15]: fixed

File: projects/cdk/directives/auto-focus/autofocus.directive.ts (hunk 3 above)
Comment (splincode) [reply projects/cdk/directives/auto-focus/autofocus.directive.ts]: fixed

File: projects/cdk/tokens/is-apple.ts (hunk 5 above)
Comment (splincode) [reply projects/cdk/tokens/is-apple.ts]: fixed

**Comment by splincode** [comment]
Unfortunately, it won’t work like this, since the state of the instance is created one time.

image

**Comment by waterplea** [comment]
> Unfortunately, it won’t work like this, since the state of the instance is created one time.
> 
> image

You need to put that in directive providers, not in module providers. Please watch our presentation on DI in Angular, Node and Module injectors:
https://www.youtube.com/watch?v=5nVKrcnIFqA

**Comment by splincode** [comment]
@waterplea sorry, I forgot basic rules, maybe I need to more sleep ((

**Comment by splincode** [comment]
@waterplea @nsbarsukov updated

File: projects/cdk/directives/auto-focus/autofocus.directive.ts [hunk 7]
```diff
@@ -1,126 +1,28 @@
...
-        @Inject(TUI_IS_IOS) private readonly isIos: boolean,
-        @Inject(ANIMATION_FRAME) private readonly animationFrame$: Observable<number>,
-        @Inject(NgZone)
-        private readonly ngZone: NgZone,
-        @Inject(ElementRef)
-        private readonly elementRef: ElementRef<HTMLElement>,
-        @Inject(Renderer2)
-        private readonly renderer: Renderer2,
+        @Inject(TUI_AUTOFOCUS_HANDLER)
+        private readonly handler: TuiAutofocusHandler | null,
```
Comment (waterplea) [inline projects/cdk/directives/auto-focus/autofocus.directive.ts:20]: Why `null`?

File: projects/cdk/utils/os/is-apple.ts [hunk 8]
```diff
@@ -0,0 +1,9 @@
+import {isIos} from './is-ios';
+
+const SAFARI_REG_EXP = /^((?!chrome|android).)*safari/i;
+
+export function isApple(navigator?: Navigator | null): boolean {
```
Comment (waterplea) [inline projects/cdk/utils/os/is-apple.ts]: Why `undefined` and `null`?

File: projects/cdk/utils/os/is-ios.ts [hunk 9]
```diff
@@ -0,0 +1,10 @@
+const IOS_REG_EXP = /ipad|iphone|ipod/;
+
+export function isIos(navigator?: Navigator | null): boolean {
```
Comment (waterplea) [inline projects/cdk/utils/os/is-ios.ts]: Same here

File: projects/cdk/directives/auto-focus/autofocus.options.ts [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+import {Observable} from 'rxjs';
+
+import {TuiDefaultAutofocusHandler} from './handlers/default.handler';
+import {TuiIosAutofocusHandler} from './handlers/ios.handler';
+
+export interface TuiAutofocusHandler {
+    setFocus(): void;
+}
+
+export const TUI_AUTOFOCUS_HANDLER = new InjectionToken<TuiAutofocusHandler | null>(
```
Comment (waterplea) [inline projects/cdk/directives/auto-focus/autofocus.options.ts:22]: Remove `null` and factory

File: projects/cdk/directives/auto-focus/autofocus.options.ts [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+// TODO: 3.0 remove in ivy compilation
+export function tuiAutofocusHandlerFactory(
+    tuiFocusableComponent: TuiFocusableElementAccessor | null,
+    elementRef: ElementRef<HTMLElement>,
+    animationFrame$: Observable<number>,
+    renderer: Renderer2,
+    ngZone: NgZone,
+): TuiAutofocusHandler {
+    return inject(TUI_IS_APPLE)
```
Comment (waterplea) [inline projects/cdk/directives/auto-focus/autofocus.options.ts:35]: I know it's more idiomatic to do it like that, but `useClass` with global `navigator` hack allows us not to care about dependencies of these classes. @nsbarsukov @vladimirpotekhin what do you think?
image

File: projects/cdk/directives/auto-focus/handlers/abstract.handler.ts [hunk 12]
```diff
@@ -0,0 +1,28 @@
+import {Directive, ElementRef} from '@angular/core';
+import {
+    TuiFocusableElementAccessor,
+    TuiNativeFocusableElement,
+} from '@taiga-ui/cdk/interfaces';
+
+import {TuiAutofocusHandler} from '../autofocus.options';
+
+@Directive()
```
Comment (waterplea) [inline projects/cdk/directives/auto-focus/handlers/abstract.handler.ts:9]: It's `@Injectable`, not a directive.

File: projects/cdk/directives/auto-focus/handlers/default.handler.ts [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+import {TUI_FOCUSABLE_ITEM_ACCESSOR} from '@taiga-ui/cdk/tokens';
+import {Observable, race, timer} from 'rxjs';
+import {map, skipWhile, take, throttleTime} from 'rxjs/operators';
+
+import {AbstractTuiAutofocusHandler} from './abstract.handler';
+
+const TIMEOUT = 1000;
+const NG_ANIMATION_SELECTOR = '.ng-animating';
+
+@Directive()
```
Comment (waterplea) [inline projects/cdk/directives/auto-focus/handlers/default.handler.ts:14]: Same

File: projects/cdk/utils/os/is-apple.ts [hunk 14]
```diff
@@ -0,0 +1,16 @@
+import {isIos} from './is-ios';
+
+const SAFARI_REG_EXP = /^((?!chrome|android).)*safari/i;
+
+export function isApple({
+    userAgent,
+    navigator,
+}: {
+    navigator: Navigator | null;
```
Comment (waterplea) [inline projects/cdk/utils/os/is-apple.ts:9]: These things are never `null` and also `userAgent` is a part of the `Navigator` so no need to inject them separately.

**Review by waterplea — CHANGES_REQUESTED** [review]

File: projects/cdk/directives/auto-focus/autofocus.options.ts (hunk 11 above)
Comment (splincode) [reply projects/cdk/directives/auto-focus/autofocus.options.ts:35]: useClass with expressions does not compile to ViewEngine

File: projects/cdk/directives/auto-focus/handlers/abstract.handler.ts (hunk 12 above)
Comment (splincode) [reply projects/cdk/directives/auto-focus/handlers/abstract.handler.ts:9]: image

File: projects/cdk/directives/auto-focus/handlers/default.handler.ts (hunk 13 above)
Comment (splincode) [reply projects/cdk/directives/auto-focus/handlers/default.handler.ts:14]: image

File: projects/cdk/directives/auto-focus/autofocus.directive.ts (hunk 7 above)
Comment (splincode) [reply projects/cdk/directives/auto-focus/autofocus.directive.ts]: fixed

File: projects/cdk/utils/os/is-apple.ts (hunk 8 above)
Comment (splincode) [reply projects/cdk/utils/os/is-apple.ts]: fixed

File: projects/cdk/utils/os/is-ios.ts (hunk 9 above)
Comment (splincode) [reply projects/cdk/utils/os/is-ios.ts]: fixed

File: projects/cdk/directives/auto-focus/autofocus.options.ts (hunk 10 above)
Comment (splincode) [reply projects/cdk/directives/auto-focus/autofocus.options.ts]: fixed

File: projects/cdk/utils/os/is-apple.ts (hunk 14 above)
Comment (splincode) [reply projects/cdk/utils/os/is-apple.ts]: fixed

**Review by waterplea — APPROVED** [review]

File: projects/cdk/directives/auto-focus/autofocus.directive.ts [hunk 15]
```diff
@@ -1,126 +1,26 @@
...
-                        map(() => this.element.closest(NG_ANIMATION_SELECTOR)),
-                        skipWhile(Boolean),
-                        take(1),
-                    ),
-                ).subscribe(() => setNativeFocused(this.element));
-            }
-
-            return;
+        if (this.autoFocus) {
+            this.handler?.setFocus();
```
[the remark this answers is not in the document]
Comment (splincode) [reply projects/cdk/directives/auto-focus/autofocus.directive.ts]: fixed

File: projects/cdk/directives/auto-focus/autofocus.directive.ts [hunk 16]
```diff
@@ -1,126 +1,26 @@
...
 
-const TIMEOUT = 1000;
-const NG_ANIMATION_SELECTOR = '.ng-animating';
+import {
+    TUI_AUTOFOCUS_HANDLER,
+    TUI_AUTOFOCUS_PROVIDERS,
+    TuiAutofocusHandler,
+} from './autofocus.options';
 
-// TODO: 3.0 change input name to tuiAutoFocus and handle empty string
```
Comment (nsbarsukov) [inline projects/cdk/directives/auto-focus/autofocus.directive.ts:26]: why do you remove the upper TODO ?

**Review by nsbarsukov — APPROVED** [review]

File: projects/cdk/directives/auto-focus/autofocus.directive.ts (hunk 16 above)
Comment (splincode) [reply projects/cdk/directives/auto-focus/autofocus.directive.ts:26]: fixed

**Pull request merged**